File: OO/simple.go
package main

import "log"

type Component struct {
	Name string
}

type Components []Component

type Bicycle struct {
	Brand string
	Model string
	Size  float32
	Components
}

func (self Component) name() string {
	return self.Name
}

func (cats Component) Speak() string {
	return cats.Name
}

func main() {
	myComponents := Components{
		Component{"frame"},
		Component{"forks"},
		Component{"shock"},
		Component{"wheels"},
	}

	bike := Bicycle{"Scott", "YPZ2", 15.5, myComponents}

	log.Print(bike.Components[0].Speak())

	log.Print(bike)
}

File: OO/simple.js
(function() {
  function Component(name) {
    this.name = name
  }

  function Bicycle(brand, model, size, components) {
    this.Brand = brand
    this.Model = model
    this.Size  = size
    this.Components = components
  }

  var myComponents = [
    new Component("frame"),
    new Component("forks"),
    new Component("shock"),
    new Component("wheels")
  ]

  Component.prototype.getName = function() {
    return this.name
  }

  Bicycle.prototype.Components = Component.prototype


  var bike = new Bicycle("Scott", "YPZ2", 15.5, myComponents)

  console.log(bike.Components[0].getName())
})()
